### PR TITLE
NPG-3845 Refactor logging mod and add OpenTelemetry layer to Explorer

### PR DESCRIPTION
**Pull request by FelipeRosa**
I made these changes while diving deeper at the explorer's code. The idea is to have a nice way of seeing tracing information to make debugging easier.

At the moment I added the OTLP exporter which can be used to export trace data to the [OTLP colletor](https://opentelemetry.io/docs/collector/) (or any other compatible backend). I used [this example](https://github.com/grafana/tempo/tree/main/example/docker-compose/otel-collector) to spin up a collector that uses [Grafana Tempo](https://grafana.com/docs/tempo/latest/) to store and query trace data through Grafana's UI.

image

### CODE REVIEW

**Comment by saibatizoku** [comment]
This looks awesome!

**Comment by cong-or** [comment]
Looks great!

**Review by cong-or — APPROVED** [review]

**Comment by minikin** [comment]
@FelipeRosa Can we merge this one?

**Review by saibatizoku — APPROVED** [review]

File: src/jormungandr/explorer/src/logging.rs [hunk 1]
```diff
@@ -129,66 +129,83 @@ impl LogSettings {
...
-        Ok((guards, self.msgs))
+        Ok(LogGuard {
+            _nonblocking_worker_guard: nonblocking_worker_guard,
+        })
     }
 
-    fn init_subscriber(&self, non_blocking: tracing_appender::non_blocking::NonBlocking) {
-        match self.config.format {
-            LogFormat::Default | LogFormat::Plain => {
+    fn init_subscriber(
```
Comment (cong-or) [inline src/jormungandr/explorer/src/logging.rs:173]: does this subscriber have global scope? just wondering how we can log everything in the explorer?

File: src/jormungandr/explorer/src/logging.rs [hunk 2]
```diff
@@ -1,122 +1,122 @@
...
-        // Worker guards that need to be held on to.
-        let mut guards = Vec::new();
+impl Default for LogOutput {
+    fn default() -> Self {
+        Self::Stderr
+    }
+}
 
+pub struct LogGuard {
+    _nonblocking_worker_guard: tracing_appender::non_blocking::WorkerGuard,
```
[the remark this answers is not in the document]
Comment (FelipeRosa) [reply src/jormungandr/explorer/src/logging.rs:82]: From what I could understand, the `tracing_appender` that is being using to write spans has this `non_blocking` writer that spawns a separate thread for buffering and executing writes. When this guard is dropped, buffered spans are flushed and the thread is terminated. So we need to hold it for as long as we are emitting spans.

File: src/jormungandr/explorer/src/logging.rs (hunk 1 above)
Comment (FelipeRosa) [reply src/jormungandr/explorer/src/logging.rs:173]: Yes. the `.init()` calls on the tracing subscriber `Registry` tries to set it as the global subscriber (https://docs.rs/tracing-subscriber/latest/tracing_subscriber/registry/struct.Registry.html#method.init).

**Comment by FelipeRosa** [comment]
> recent changes look good +1
> 
> Only comment in relation to logging/telemetry out of the box.
> 
> do we have to add spans etc or is it good to go as is?

@saibatizoku this PR only sets up the functionally for exporting tracing events to an OpenTelemetry Collector. We still need to work on refactoring/improving the span structure and coverage.

PS: Exporting is disabled if no Collector endpoint is set while starting the explorer so things should work as before in that way.

**Pull request merged**